File: dispatch/cli/src/protocol/primitives.ts
// Dispatch protocol primitives.
//
// These types define the universal vocabulary for AI session observability.
// Every adapter — regardless of the backend (Claude Code, Codex, Gemini,
// Ollama, a browser extension, etc.) — maps its native events onto these
// primitives.  The phone app renders these and nothing else.
//
// Aligned with Vercel AI SDK LanguageModelV3 content types where possible,
// but transport-agnostic and designed for streaming over encrypted WebSocket
// rather than HTTP/SSE.

// ---------------------------------------------------------------------------
// Session — a long-lived connection to one agent
// ---------------------------------------------------------------------------

export type SessionStatus = "connecting" | "active" | "idle" | "error" | "closed";

export interface Session {
  /** Unique session identifier (UUID). */
  id: string;
  /** Human-readable label (e.g. "Claude Code — dispatch repo"). */
  name: string;
  /** Adapter type that produced this session. */
  adapterType: string;
  /** Current lifecycle status. */
  status: SessionStatus;
  /** Working directory on the host machine, if applicable. */
  cwd?: string;
  /** Model identifier, if known (e.g. "claude-sonnet-4-20250514"). */
  model?: string;
  /** Provider-specific metadata (AI SDK providerOptions pattern). */
  providerMeta?: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Turn — one request/response cycle within a session
// ---------------------------------------------------------------------------

export type TurnStatus = "started" | "streaming" | "completed" | "failed" | "stopped";

export interface Turn {
  /** Unique turn identifier. */
  id: string;
  /** Parent session. */
  sessionId: string;
  /** Lifecycle status. */
  status: TurnStatus;
  /** ISO-8601 timestamp when the turn started. */
  startedAt: string;
  /** ISO-8601 timestamp when the turn reached a terminal state. */
  endedAt?: string;
  /** Ordered sequence of content blocks produced during this turn. */
  blocks: Block[];
}

// ---------------------------------------------------------------------------
// Block — a discrete unit of content within a turn
//
// Aligned with AI SDK V3 content types:
//   text       → LanguageModelV3TextContent
//   reasoning  → LanguageModelV3ReasoningContent
//   action     → LanguageModelV3ToolCall + ToolResult (unified)
//   file       → LanguageModelV3FileContent
//   error      → (no direct equivalent — Dispatch addition)
// ---------------------------------------------------------------------------

export type BlockStatus = "started" | "streaming" | "completed" | "failed";

export type Block = TextBlock | ReasoningBlock | ActionBlock | FileBlock | ErrorBlock;

interface BlockBase {
  /** Unique block identifier within the turn. */
  id: string;
  /** Parent turn. */
  turnId: string;
  /** Lifecycle status. */
  status: BlockStatus;
  /** Monotonic index for stable ordering. */
  index: number;
}

/** Free-form text from the agent (markdown). */
export interface TextBlock extends BlockBase {
  type: "text";
  text: string;
}

/** Agent's internal reasoning / chain-of-thought (may be collapsed in UI). */
export interface ReasoningBlock extends BlockBase {
  type: "reasoning";
  text: string;
}

/**
 * The agent performing an action.  The `kind` field carries structured
 * metadata so the client can render specialized views (diff viewer for
 * file_change, terminal for command, generic card for unknown tools).
 *
 * Aligns with AI SDK V3 tool-call + tool-result, but unified into a single
 * block with lifecycle tracking rather than two separate content parts.
 */
export interface ActionBlock extends BlockBase {
  type: "action";
  action: Action;
}

/** File content produced by the agent (images, generated files). */
export interface FileBlock extends BlockBase {
  type: "file";
  /** MIME type (e.g. "image/png", "text/plain"). */
  mimeType: string;
  /** File name or path, if known. */
  name?: string;
  /** Base64-encoded data, or a URI reference. */
  data: string;
}

/** Something went wrong during this block. */
export interface ErrorBlock extends BlockBase {
  type: "error";
  message: string;
  code?: string;
}

// ---------------------------------------------------------------------------
// Action variants — the `kind` discriminator on ActionBlock
// ---------------------------------------------------------------------------

export type Action =
  | FileChangeAction
  | CommandAction
  | ToolCallAction
  | SubagentAction;

interface ActionBase {
  /** Current status of the action. */
  status: "pending" | "running" | "completed" | "failed" | "awaiting_approval";
  /** Streaming output text (terminal output, diff chunks, etc.). */
  output: string;

  /** Present when status is "awaiting_approval". */
  approval?: {
    /** Monotonic version — incremented each time this action's approval
     *  state changes. Prevents stale phone responses from taking effect. */
    version: number;
    /** Human-readable description of what's being requested. */
    description?: string;
    /** Risk level hint for UI treatment (color, prominence). */
    risk?: "low" | "medium" | "high";
  };
}

/** Agent editing a file (maps to AI SDK file-change tool pattern). */
export interface FileChangeAction extends ActionBase {
  kind: "file_change";
  path: string;
  diff?: string;
}

/** Agent executing a shell command. */
export interface CommandAction extends ActionBase {
  kind: "command";
  command: string;
  exitCode?: number;
}

/** Generic tool invocation (AI SDK V3 tool-call). */
export interface ToolCallAction extends ActionBase {
  kind: "tool_call";
  toolName: string;
  toolCallId: string;
  input?: unknown;
  result?: unknown;
}

/** Agent spawning or coordinating with a sub-agent. */
export interface SubagentAction extends ActionBase {
  kind: "subagent";
  agentId: string;
  agentName?: string;
  prompt?: string;
}

// ---------------------------------------------------------------------------
// Delta — a streaming update to a block
//
// Mirrors the AI SDK V3 stream part lifecycle:
//   block:start  → text-start / reasoning-start / tool-input-start
//   block:delta  → text-delta / reasoning-delta / tool-input-delta
//   block:end    → text-end / reasoning-end / tool-output-available
// ---------------------------------------------------------------------------

export type Delta =
  | BlockStartDelta
  | BlockTextDelta
  | BlockActionOutputDelta
  | BlockActionStatusDelta
  | BlockActionApprovalDelta
  | BlockEndDelta;

/** A new block has started within a turn. */
export interface BlockStartDelta {
  event: "block:start";
  sessionId: string;
  turnId: string;
  block: Block;
}

/** Append text to a text or reasoning block. */
export interface BlockTextDelta {
  event: "block:delta";
  sessionId: string;
  turnId: string;
  blockId: string;
  text: string;
}

/** Append output to an action block (terminal output, diff chunk, etc.). */
export interface BlockActionOutputDelta {
  event: "block:action:output";
  sessionId: string;
  turnId: string;
  blockId: string;
  output: string;
}

/** Action status changed (pending → running → completed/failed). */
export interface BlockActionStatusDelta {
  event: "block:action:status";
  sessionId: string;
  turnId: string;
  blockId: string;
  status: Action["status"];
  /** Set on completion — exit code, result, etc. */
  meta?: Record<string, unknown>;
}

/** Action transitioned to awaiting_approval — phone renders approve/deny UI. */
export interface BlockActionApprovalDelta {
  event: "block:action:approval";
  sessionId: string;
  turnId: string;
  blockId: string;
  approval: {
    version: number;
    description?: string;
    risk?: "low" | "medium" | "high";
  };
}

/** A block has reached a terminal state. */
export interface BlockEndDelta {
  event: "block:end";
  sessionId: string;
  turnId: string;
  blockId: string;
  status: BlockStatus;
}

// ---------------------------------------------------------------------------
// Turn lifecycle events
// ---------------------------------------------------------------------------

export type TurnEvent =
  | { event: "turn:start"; sessionId: string; turn: Turn }
  | { event: "turn:end"; sessionId: string; turnId: string; status: TurnStatus }
  | { event: "turn:error"; sessionId: string; turnId: string; message: string };

// ---------------------------------------------------------------------------
// Session lifecycle events
// ---------------------------------------------------------------------------

export type SessionEvent =
  | { event: "session:update"; session: Session }
  | { event: "session:closed"; sessionId: string };

// ---------------------------------------------------------------------------
// Wire message — the union of everything that flows over the encrypted pipe
// ---------------------------------------------------------------------------

export type DispatchEvent = Delta | TurnEvent | SessionEvent;

// ---------------------------------------------------------------------------
// Prompt — user input going into a session
// ---------------------------------------------------------------------------

export interface Prompt {
  /** Target session. */
  sessionId: string;
  /** Text content. */
  text: string;
  /** File paths mentioned / attached. */
  files?: string[];
  /** Images attached (base64). */
  images?: Array<{ mimeType: string; data: string }>;
  /** Provider-specific options pass-through. */
  providerOptions?: Record<string, unknown>;
}

File: scout/ios/Scout/Models/Primitives.swift
// Primitives — Dispatch protocol types.
//
// Canonical source: PROTOCOL.md §5.
// These types mirror the TypeScript definitions in src/protocol/primitives.ts
// and must stay in sync with the wire format.

import Foundation

// MARK: - Session

enum SessionStatus: String, Codable, Sendable {
    case connecting
    case active
    case idle
    case error
    case closed
}

struct Session: Codable, Identifiable, Sendable {
    let id: String
    var name: String
    let adapterType: String
    var status: SessionStatus
    var cwd: String?
    var model: String?
    var providerMeta: [String: AnyCodable]?
}

// MARK: - Turn

enum TurnStatus: String, Codable, Sendable {
    case started
    case streaming
    case completed
    case failed
    case stopped
}

struct Turn: Codable, Identifiable, Sendable {
    let id: String
    let sessionId: String
    var status: TurnStatus
    let startedAt: String
    var endedAt: String?
    var blocks: [Block]
    var isUserTurn: Bool?
    var turnHash: String? = nil
}

// MARK: - Block

enum BlockStatus: String, Codable, Sendable {
    case started
    case streaming
    case completed
    case failed
}

struct Block: Codable, Identifiable, Sendable {
    let id: String
    let turnId: String
    let type: BlockType
    var status: BlockStatus
    let index: Int

    // Text / reasoning fields
    var text: String?

    // Action fields
    var action: Action?

    // File fields
    var mimeType: String?
    var name: String?
    var data: String?

    // Error fields
    var message: String?
    var code: String?
}

enum BlockType: String, Codable, Sendable {
    case text
    case reasoning
    case action
    case file
    case error
}

// MARK: - Action

enum ActionKind: String, Codable, Sendable {
    case fileChange = "file_change"
    case command
    case toolCall = "tool_call"
    case subagent
}

enum ActionStatus: String, Codable, Sendable {
    case pending
    case running
    case completed
    case failed
}

struct Action: Codable, Sendable {
    let kind: ActionKind
    var status: ActionStatus
    var output: String

    // file_change
    var path: String?
    var diff: String?

    // command
    var command: String?
    var exitCode: Int?

    // tool_call
    var toolName: String?
    var toolCallId: String?
    var input: AnyCodable?
    var result: AnyCodable?

    // subagent
    var agentId: String?
    var agentName: String?
    var prompt: String?
}

// MARK: - Prompt

struct Prompt: Codable, Sendable {
    let sessionId: String
    var text: String
    var files: [String]?
    var images: [ImageAttachment]?
    var providerOptions: [String: AnyCodable]?
}

struct ImageAttachment: Codable, Sendable {
    let mimeType: String
    let data: String
}

// MARK: - Snapshot types

struct SessionState: Codable, Sendable {
    var session: Session
    var history: SessionHistory?
    var turns: [TurnState]
    var currentTurnId: String?
}

struct SessionHistory: Codable, Sendable {
    var hasOlder: Bool
    var oldestTurnId: String?
    var newestTurnId: String?
}

enum SnapshotTurnStatus: String, Codable, Sendable {
    case streaming
    case completed
    case interrupted
    case error
}

struct TurnState: Codable, Sendable {
    let id: String
    var status: SnapshotTurnStatus
    var blocks: [BlockState]
    let startedAt: Int
    var endedAt: Int?
    var isUserTurn: Bool?
    var turnHash: String? = nil
}

enum SnapshotBlockStatus: String, Codable, Sendable {
    case streaming
    case completed
}

struct BlockState: Codable, Sendable {
    var block: Block
    var status: SnapshotBlockStatus
}

struct SessionSummary: Codable, Identifiable, Sendable {
    let sessionId: String
    var name: String
    let adapterType: String
    var status: String
    var turnCount: Int
    var currentTurnStatus: String?
    let startedAt: Int
    var lastActivityAt: Int
    var project: String? = nil
    var model: String? = nil
    var isCachedOnly: Bool = false

    var id: String { sessionId }
}

// MARK: - AnyCodable (type-erased JSON)

struct AnyCodable: Codable, @unchecked Sendable {
    let value: Any

    init(_ value: Any) {
        self.value = value
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            value = NSNull()
        } else if let bool = try? container.decode(Bool.self) {
            value = bool
        } else if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let string = try? container.decode(String.self) {
            value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            value = array.map(\.value)
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            value = dict.mapValues(\.value)
        } else {
            value = NSNull()
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch value {
        case is NSNull:
            try container.encodeNil()
        case let bool as Bool:
            try container.encode(bool)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let array as [Any]:
            try container.encode(array.map { AnyCodable($0) })
        case let dict as [String: Any]:
            try container.encode(dict.mapValues { AnyCodable($0) })
        default:
            try container.encodeNil()
        }
    }
}

// MARK: - Typed helpers

extension AnyCodable {
    var stringValue: String? {
        value as? String
    }
}

extension Session {
    var agentId: String? {
        providerMeta?["agentId"]?.stringValue?.trimmedNonEmpty
    }

    var currentBranch: String? {
        for key in ["branch", "gitBranch", "currentBranch", "workspaceQualifier"] {
            if let branch = providerMeta?[key]?.stringValue?.trimmedNonEmpty {
                return branch
            }
        }

        if let selector = providerMeta?["selector"]?.stringValue?.trimmedNonEmpty,
           let suffix = selector.split(separator: "/").last,
           !suffix.isEmpty {
            return String(suffix)
        }

        return nil
    }

    var inferredProjectName: String? {
        if let explicitProject = providerMeta?["project"]?.stringValue?.trimmedNonEmpty {
            return explicitProject
        }

        if let cwd = cwd?.trimmedNonEmpty {
            let lastComponent = URL(fileURLWithPath: cwd).lastPathComponent.trimmingCharacters(in: .whitespacesAndNewlines)
            if !lastComponent.isEmpty && lastComponent != "/" {
                return lastComponent
            }
        }

        for separator in [" — ", " – ", " · "] {
            let parts = name.components(separatedBy: separator)
            if let suffix = parts.last?.trimmingCharacters(in: .whitespacesAndNewlines),
               !suffix.isEmpty,
               suffix != name {
                return suffix
            }
        }

        return nil
    }
}

extension String {
    var trimmedNonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}
